Upgrade faer to the latest version

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit updates faer to the latest version 0.23.2. This also requires an update to faer-ext as the two crates are tightly coupled. The faer-ext version bump does depend on the latest nalgebra release when the nalgebra conversion feature is enabled. However, the latest nalgebra release requires a newer rust version than our MSRV and we can't upgrade to it. To workaround this issue, this commit adds conversion functions to convert between the types by reference with no copies. For the faer->nalgebra this is basically just an inlining of what faer-ext does, but for nalgebra->faer it goes through ndarray (since we already have that).

### Details and comments

mtreinish committed:
Upgrade faer to the latest version

This commit updates faer to the latest version 0.23.1. This also
requires an update to faer-ext as the two crates are tightly coupled.
The faer-ext version bump does depend on the latest nalgebra release
when the nalgebra conversion feature is enabled. However, the latest
nalgebra release requires a newer rust version than our MSRV and we
can't upgrade to it. To workaround this issue, this commit adds
conversion functions to convert between the types by reference with no
copies. For the faer->nalgebra this is basically just an inlining of
what faer-ext does, but for nalgebra->faer it goes through ndarray
(since we already have that).

mtreinish commented:
Sigh, it looks like the latest version is still broken on x86_64 macOS and isn't compiling. We'll have to wait for a fix before we can upgrade this.

mtreinish committed:
Bump to newer bugfix

mtreinish committed:
Bump private-gemm-x86

mtreinish committed:
Bump private-gemm-x86 again

mtreinish committed:
Merge remote-tracking branch 'origin/main' into update-aer

mtreinish commented:
I was curious and this seems to mostly fix the regression introduced in #14873 :

```
Benchmarks that have improved:

| Change   | Before [ed4b04f6] <update-aer^2>   | After [9b4fcbce] <update-aer>   |   Ratio | Benchmark (Parameter)                                             |
|----------|------------------------------------|---------------------------------|---------|-------------------------------------------------------------------|
| -        | 77.0±1ms                           | 69.9±1ms                        |    0.91 | utility_scale.UtilityScaleBenchmarks.time_bv_100('cx')            |
| -        | 85.9±2ms                           | 75.9±0.6ms                      |    0.88 | utility_scale.UtilityScaleBenchmarks.time_bv_100('cz')            |
| -        | 599±60ms                           | 514±20ms                        |    0.86 | utility_scale.UtilityScaleBenchmarks.time_qft('cz')               |
| -        | 115±4ms                            | 97.8±0.8ms                      |    0.85 | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cz') |
| -        | 354±10ms                           | 283±5ms                         |    0.8  | utility_scale.UtilityScaleBenchmarks.time_qaoa('cz')              |

Benchmarks that have stayed the same:

| Change   | Before [ed4b04f6] <update-aer^2>   | After [9b4fcbce] <update-aer>   | Ratio   | Benchmark (Parameter)                                                         |
|----------|------------------------------------|---------------------------------|---------|-------------------------------------------------------------------------------|
|          | 615±30ms                           | 558±20ms                        | ~0.91   | utility_scale.UtilityScaleBenchmarks.time_qv('ecr')                           |
|          | 491±30ms                           | 443±10ms                        | ~0.90   | utility_scale.UtilityScaleBenchmarks.time_qv('cx')                            |
|          | 79.3±4ms                           | 71.0±2ms                        | ~0.90   | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cx')             |
|          | 0                                  | 0                               | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cx')                 |
|          | 0                                  | 0                               | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cz')                 |
|          | 0                                  | 0                               | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('ecr')                |
|          | 6.83±0.03s                         | 7.27±0.09s                      | 1.06    | utility_scale.UtilityScaleBenchmarks.time_circSU2('ecr')                      |
|          | 87.6±5ms                           | 91.7±0.4ms                      | 1.05    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cz')                |
|          | 65.3±0.4ms                         | 68.0±0.4ms                      | 1.04    | utility_scale.UtilityScaleBenchmarks.time_bv_100('ecr')                       |
|          | 6.06±0.3ms                         | 6.14±0.3ms                      | 1.01    | utility_scale.UtilityScaleBenchmarks.time_bvlike('cz')                        |
|          | 90.2±4ms                           | 91.1±2ms                        | 1.01    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cx')                |
|          | 6.83±0.02s                         | 6.82±0.02s                      | 1.00    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cx')                       |
|          | 6.83±0s                            | 6.85±0.01s                      | 1.00    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cz')                       |
|          | 400                                | 400                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cx')                 |
|          | 400                                | 400                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cz')                 |
|          | 400                                | 400                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('ecr')                |
|          | 300                                | 300                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cx')                |
|          | 300                                | 300                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cz')                |
|          | 300                                | 300                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('ecr')               |
|          | 1590                               | 1590                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cx')                   |
|          | 1603                               | 1603                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cz')                   |
|          | 1603                               | 1603                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('ecr')                  |
|          | 2692                               | 2692                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cx')                    |
|          | 2744                               | 2744                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cz')                    |
|          | 2744                               | 2744                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('ecr')                   |
|          | 2571                               | 2571                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cx')                     |
|          | 2571                               | 2571                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cz')                     |
|          | 2571                               | 2571                            | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('ecr')                    |
|          | 480                                | 480                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cx')      |
|          | 480                                | 480                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cz')      |
|          | 480                                | 480                             | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('ecr')     |
|          | 27.9±0.2ms                         | 27.5±1ms                        | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('ecr') |
|          | 429±20ms                           | 427±9ms                         | 0.99    | utility_scale.UtilityScaleBenchmarks.time_qft('ecr')                          |
|          | 29.1±1ms                           | 28.5±1ms                        | 0.98    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cx')  |
|          | 28.8±1ms                           | 28.2±1ms                        | 0.98    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cz')  |
|          | 6.22±0.09ms                        | 6.00±0.2ms                      | 0.96    | utility_scale.UtilityScaleBenchmarks.time_bvlike('ecr')                       |
|          | 164±3ms                            | 158±2ms                         | 0.96    | utility_scale.UtilityScaleBenchmarks.time_qaoa('cx')                          |
|          | 90.8±1ms                           | 86.1±4ms                        | 0.95    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('ecr')               |
|          | 85.8±0.7ms                         | 81.8±2ms                        | 0.95    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('ecr')            |
|          | 6.18±0.2ms                         | 5.79±0.2ms                      | 0.94    | utility_scale.UtilityScaleBenchmarks.time_bvlike('cx')                        |
|          | 8.31±0.3ms                         | 7.79±0.04ms                     | 0.94    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cx')               |
|          | 8.40±0.4ms                         | 7.90±0.3ms                      | 0.94    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cz')               |
|          | 296±8ms                            | 279±6ms                         | 0.94    | utility_scale.UtilityScaleBenchmarks.time_qft('cx')                           |
|          | 217±9ms                            | 203±6ms                         | 0.93    | utility_scale.UtilityScaleBenchmarks.time_qaoa('ecr')                         |
|          | 8.66±0.1ms                         | 7.97±0.3ms                      | 0.92    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('ecr')              |
|          | 604±10ms                           | 558±40ms                        | 0.92    | utility_scale.UtilityScaleBenchmarks.time_qv('cz')                            |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -12,15 +12,60 @@
+    let dim = ::ndarray::Dim(mat.shape());
+    let strides = ::ndarray::Dim(mat.strides());
+
+    // SAFETY: We know the array is a 2x2 and contiguous block (DMatrix uses a vec for backing        storage) so we
```
How do we know it's 2x2? Did you mean 2d?

mtreinish replied:
Sigh, yeah I meant 2d. I copied this from a different place where I did something similar but where it was a `Matrix2` not a `DMatrix`.

```suggestion
    // SAFETY: We know the array is a 2d array in a contiguous block (DMatrix uses a vec for backing storage) so we
```

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -12,15 +12,60 @@
+    // so creating a view of the data from the from the pointer is safe unless the faer object
+    // is already corrupt.
+    unsafe {
+        MatrixView::<'_, Complex64, Dyn, Dyn, Dyn, Dyn>::from_data(ViewStorage::<
```
This is just a note to myself to look at this in more detail. (I'm on my phone right now.)

jakelishman commented on Cargo.lock:
```diff
@@ -512,6 +571,12 @@ version = "2.3.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "37909eebbb50d72f9059c3b6d82c0463f2ff062c9e95845c43a6c9c0355411be"
 
+[[package]]
+name = "find-msvc-tools"
```
Were we planning to do anything about all the Windows dependencies?

mtreinish replied:
We can ask `faer` about it. But the current state of this version is that to enable rayon it pulls in `spindle` which pulls in `loom` which pulls in all these dependencies. `loom` based on the [description](https://crates.io/crates/loom) seems like an odd fit to have as a hard dependency. If we want to exclude these dependencies from the lock file our only option would be to disable rayon for faer usage (it's in the default features). I think prior to #14797 that might have been more viable but rayon actively is being used now since we are working with larger matrices.

jakelishman reviewed (commented):
I need to come back to look at the unsafe code again, this was just a couple of small questions.

mtreinish committed:
Update crates/synthesis/src/linalg/mod.rs

mtreinish committed:
Merge branch 'main' into update-aer

mtreinish committed:
Merge branch 'main' into update-aer

mtreinish committed:
Merge remote-tracking branch 'origin/main' into update-aer

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -12,15 +12,60 @@
+            (Dyn(nrows), Dyn(ncols)),
+            (
+                Dyn(row_stride.try_into().unwrap()),
+                Dyn(col_stride.try_into().unwrap()),
```
`row_stride` and `col_stride` return `isize`, which is why the `try_into` is necessary, I think?

Can we put an `expect("only works for positive strides")` or a comment in the code / something to document the requirement?  I'm particularly worried that we might accidentally come up against this if we take the transpose of a matrix before taking out an `nalgebra` view.

The documentation of `ViewStorage` says that it requires the backing data to be `[Complex64; R * C]` for safety, which we don't assert here, and I'm kind of assuming is possible to violate from Faer by producing a strided view onto an array:

1. do we know if it's just a documentation bug of `nalgebra`? or
2. is it actually possible to produce a non-contiguous backing view in faer? or
3. should we put a contiguous assertion somewhere in the code before we get here

mtreinish replied:
Part of me was acting on trust here; most of this implementation was taken from the faer conversion trait's implementation (hence the attribution comment) and I was assuming it was safe. To answer your questions

1. I'd have to dive into the code to double check what the underlying assumptions are on `ViewStorage` are (it. I know that nalgebra supports strided views: https://docs.rs/nalgebra/latest/nalgebra/base/struct.Matrix.html#method.view_with_steps so I think the docs have to be wrong or at least misleading because you could totally end up a view that is accessing outside `[T; R * C]` so it has to be something like at least that big, but more depending on the striding. 
2.  I think in general it is definitely possible. But in how it's used right now it is not. The only caller of this function right now is from the return of running svd. The methods we call it on are just returning views of the owned contiguous faer matrices in the svd result object. However, I expect in the future (e.g. with #15017) there will be a need to use this in more places.
3. I think this is actually valid usage despite the docs, like nalgebra supports strided views on the array so even if faer was returning a strided `MatRef` so I don't think a contiguous assertion is correct. If you want an assertion here I think the only one that would make sense is that the number of elements in the backing storage is at least R*C in size, but an exact equality can't be right.

mtreinish replied:
I switched to expect in: https://github.com/Qiskit/qiskit/pull/15059/commits/13c2f13f6a301b20d17acf034f549378a6f399c6

jakelishman replied:
The contiguous assertion was just for the case that we're actually doing something wrong by producing a strided view.  If it's valid, then no need to assert it away.

Looking at the code for `ViewStorage::into_slice` (which is what _definitely_ drives the requirements), I'm pretty convinced it's a documentation bug in `from_raw_parts`: https://docs.rs/nalgebra/0.34.1/src/nalgebra/base/matrix_view.rs.html#106-120.  The code constructs a slice that's dependent on the strides as well, so it can include more elements.  Technically, there's the possibility for undefined behaviour here still, though - if we were able to safely produce a strided array in Faer somehow that strides _over_ (without making accessible) uninitialised memory within the pointer, and we then called this function, the caller, through entirely safe code, could then do `into_slice` and cause UB by having a reference to uninitialised memory.

I haven't looked in detail into the Faer code to see if that contiguous/initialised assumption is one we definitely can safely make.  _Probably_ I guess we can because Faer's already doing it?

> I think in general it is definitely possible. But in how it's used right now it is not.

I know you go on to say "but let's take care because we might use it" right after this, but I still don't like the words "how it's used right now" being anywhere _near_ discussions about `unsafe` haha

alexanderivrii replied:
Would it make sense to add rust tests for these conversions where we would construct a faer (resp. nalgebra) matrix, take the transpose and/or the submatrix, and convert to nalgebra (resp. faer)?

mtreinish replied:
I can add tests, it doesn't hurt and would give us some coverage via miri. But realistically I think the concern Jake is worried about is probably something we would actively have to try to do in our typical rust usage. The only place we use faer in Qiskit is to run its linalg functions and we only call this conversion in places where we generate a matrix from a linalg decomposition function in faer and we need to work with that in nalgebra for further analysis. In these cases this won't come up because faer is going to allocate a contiguous backing array to work with, even if it returns a strided view of the matrix it will be in a larger contiguous allocation. For us to trigger this we'll have to accept an arbitrary input array and build a `MatRef` from it which exhibits this behavior and then also run `into_slice()` after the conversion (which I don't think is something we ever do). The only path for this would be explicitly doing it in unsafe rust code, writing a python function that takes in a numpy array setup like this by reference, converts it to an `ArrayView2` then make a `MatRef` from that. The array inputs in the C API explicitly assert the input has to be a contiguous and we also copy the contents into a contiguous array too.

But Jake is correct this doesn't stop someone from trying to use this incorrectly with an arbitrary faer `MatRef` which could be back with non contiguous memory strided around uninitialized memory. They would have to make the function public first since this PR has this function as private to this module which is safe. But as I said in my earlier comment I expect we'll want to use this in the two qubit decomposer before too long as well so we will likely make it public in the near future. From my perspective this doesn't really make us anymore unsafe than what is there now, it's just hiding in faer-ext which is where this conversion logic came from. I guess I can document this concern in a docstring in anticipation this function will be public in a subsequent PR.

TL;DR I'll add some tests just to have direct coverage, and also add a docstring. I can also make the function `unsafe` to make people assert they know what they're doing here.

mtreinish replied:
I added docs and tests in: https://github.com/Qiskit/qiskit/pull/15059/commits/91284ea8e614e2f93ebe6860284efcaad34b5537

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -12,15 +12,60 @@
+    // SAFETY: We know the array is a 2d array in a contiguous block (DMatrix uses a vec for backing storage) so we
+    // don't need to check for invalid format
+    let array = unsafe { ArrayView2::from_shape_ptr(dim.strides(strides), mat.get_unchecked(0)) };
+    array.into_faer()
```
Similar comment about strides here: are we certain that you can't produce a strided view?  The _owned_ object might be contiguous, but are we 100% certain that all views are, and is it a problem if they aren't?

mtreinish replied:
That's fair I copied this comment (and the bulk of this function code) from a context where it is an owned matrix type which is allocated without striding going to an `ArrayView2`. But the typing here would allow any type of view, it should still work fine even with that because we are telling the details of the striding to ndarray. I'll update the safety comment.

mtreinish replied:
I've updated the safety comment in: https://github.com/Qiskit/qiskit/pull/15059/commits/5a4c4bf8c591869a477cdd639c903fdbff31452c

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -12,15 +12,60 @@
 
+fn nalgebra_to_faer<R: Dim, C: Dim, RStride: Dim, CStride: Dim>(
+    mat: MatrixView<Complex64, R, C, RStride, CStride>,
+) -> MatRef<Complex64> {
```
As a minor point: can you make the elided but linked lifetimes explicit in this and the other function, or will clippy complain about it?

It'd be nice to have it super explicit for these hairy functions, but it's not a big deal if the linter will moan.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/15059/commits/87c2f94156625570b778f5e7e389185275f64217

mtreinish committed:
Add explicit lifetime parameters to function signature

alexanderivrii commented:
Just a silly comment that the summary says "This commit updates faer to the latest version 0.23.1" while the code actually updates it to 0.23.2 (which granted was updated since after the PR was first opened).

mtreinish committed:
Update safety comment for nalgebra_to_faer

mtreinish commented:
> Just a silly comment that the summary says "This commit updates faer to the latest version 0.23.1" while the code actually updates it to 0.23.2 (which granted was updated since after the PR was first opened).

I've updated it in the PR summary, that was the text in the[ first commit message](https://github.com/Qiskit/qiskit/pull/15059/commits/917fc409506c4456404295f95b2ace093bf9a5b7) on the branch and as you said it was 0.23.1 at the time.

mtreinish committed:
Update stride type conversion panic message

mtreinish committed:
Run cargo fmt

mtreinish committed:
Fix clippy failures too

mtreinish committed:
Add tests and docs for faer conversion function

There are some caveats around using the faer conversion function around
how nalgebra expects the memory layout to be setup. This documents them
to make it clear there is some care needed with the function.
Additionally, this commit adds some test coverage to the function so that
we are directly covering it and potentially running it under miri (I'm
not actually sure if this will work under miri, so we'll find out in CI
and go from there).

mtreinish committed:
Merge branch 'main' into update-aer

mtreinish committed:
Test something besides an identity matrix

mtreinish committed:
Merge remote-tracking branch 'origin/main' into update-aer

alexanderivrii reviewed (dismissed):
Thanks for explaining the safety concerns and adding the tests for normal usage flows. I am marking it as approved, but will delegate merging to @jakelishman.

mtreinish committed:
Skip miri on test_transpose_faer_to_nalgebra_conversion

This commit adds a skip on the
test_transpose_faer_to_nalgebra_conversion test when running under miri.
Miri is flagging undefined behavior in the nalgebra code which is
triggered by advancing the iterator's pointer outside the allocation.
This is because the nalgebra code is using the wrong pointer api for doing
the pointer addition. A PR has been proposed in dimforge/nalgebra#1562 to
fix this, but until that is merged and in a release Qiskit is using we
will need to skip miri on this test.

mtreinish committed:
Merge branch 'main' into update-aer

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -128,3 +192,42 @@ pub fn svd_decomposition(
+        assert_eq!(mat_view, expected);
+    }
+
+    #[cfg(not(miri))] // TODO: Remove this after dimforge/nalgebra#1562 is released
```
At this point we'll be able to relax some of the "warning" comment in the conversion method too, I think?

mtreinish replied:
I think the warning is still valid, there are definitely edge cases in the nalgebra code around `MatrixView` and warning people about it still seems prudent. While the common case is fine I think, there are definitely potential edge cases if you're doing something weird.

jakelishman commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -12,15 +12,79 @@
 const ATOL_DEFAULT: f64 = 1e-8;
 const RTOL_DEFAULT: f64 = 1e-5;
 
+fn nalgebra_to_faer<R: Dim, C: Dim, RStride: Dim, CStride: Dim>(
```
Should this (and `faer_to_nalgebra`) be more public than this, or are we deliberately limiting scope?

mtreinish replied:
I was limiting scope for now, I was thinking we could make this public in a follow-up when we need it and also make these generic over `T` at that time. For right now this is the only place we do the direct conversions so I wanted to keep it local just for the faer upgrade.

jakelishman replied:
That's fine by me 👍

jakelishman reviewed (approved):
In principle I think this is ok now, thanks for the effort in wading through the `unsafe` problems.

There's a lot of sharp edges, but I think we're in a place where we've got a handle on them, have documented them to ourselves, and can hopefully work with `nalgebra` and `faer` to clean some of them out if necessary.  I'm ok to merge and go from here - feel free to enqueue if the comments don't need any action.
